# BlueSky Starter Pack Data Collection

Team name: A5

Team members:
- **Trang Kieu:** Data collection and code review
- **Terresa Tran:** Helped along with Trang and Mei with coding and result interpretations
- **Wynne Tseng:** Report analysis and results
- **Mei Wu:** Did thematic categorization code
- **Vivien Wang:** Review and orgainzation of report

### Q1: Qualitative: Concept Definition, Context, and Rationale


#### Context and Motivation

Bluesky starter packs are curated lists of accounts and feeds (up to 150 people and up to 3 custom feeds) designed to help users discover communities and content when joining the platform. Because Bluesky is decentralized and does not rely on traditional algorithmic recommendation systems to guide content discovery, users must depend more on manual or curated tools to navigate the network. As a result,  starter packs play a critical role in shaping user discovery and social network formation.

According to "Bootstrapping Social Networks: Lessons from Bluesky Starter Packs", starter packs often circulate within communities, creating clusters or "social bubbles" where users promote and follow others within the same social groups. This suggests that starter packs may reinforce community structures and amplify visibility of certain accounts.

Understanding these patterns helps answer broader questions about influence, discovery, and community formation in decentralized social networks.

#### Concept Definition: Echo Chamber

An echo chamber refers to a social environment in which users are primarily exposed to information, accounts, and viewpoints that align with their existing perspectives, while alternative perspectives remain marginal or absent.

In the context of Bluesky starter packs, echo chambers may form when starter packs repeatedly recommend accounts that are highly similar in terms of social connections, interests, views, or topical focus. This can lead to clustered communities where information circulates within the same group.

Because starter packs function as curated recommendation systems, they tend to introduce users to specific communities rather than broader and more diverse networks, it may contribute to echo chambers that users could become embedded in relatively homogenous clusters, reinforcing existing perspectives over time. 

##### We are going to define echo chamber through:
Follower overlap percentage between accounts in the same pack


#### Hypothesis

There are shared patterns across Bluesky starter packs. Specifically:
Some accounts appear frequently across many starter packs, indicating higher visibility or influence within the network.
Certain feeds are repeatedly included, suggesting they play a central role in content discovery.
Starter packs reflect community clusters or "social bubbles," in which users promote accounts within their own communities.

#### Research Questions

This phase focuses on four primary questions:

RQ1: Which accounts appear most frequently across starter packs?  
RQ2: Which feeds appear most frequently across starter packs?  
RQ3: Do starter pack descriptions suggest themes (art, politics, tech, sports, etc.)?  
RQ4: Do certain creators create many starter packs?

Starter packs may be organized around shared interests or communities.

#### Assumptions and Potential Biases
- Starter packs represent intentional curation and community knowledge
- Frequency of inclusion approximates visibility/importance
- Starter packs may overrepresent certain communities and underrepresent others depending on which URIs we could collect

### Constraints:
The Bluesky API returns details for a starter pack only when given a specific URI, so our approach depends on first collecting a list of starter pack URIs, then fetching each pack individually.

Also, for phase 1 of this project, we are just going to analyze 1,000 randomnized starter packs to see if our hypothesis is correct first, as there is 300,000 starter packs to analyze.

In [83]:
#!pip install atproto --quiet

In [84]:
# Import Libraries
import json

from atproto import Client, models
from atproto import exceptions
from password import BSKY_USERNAME, BSKY_APP_PASSWORD
import pandas as pd
from typing import List, Dict



In [85]:
# Enter your Bluesky Username and password for authentication
# Note: You can also create a file name password in the same directory and then store your user name as BSKY_USERNAME and password as BSKY_APP_PASSWORD.
# This Jupyter Notebook will import password file and your BSKY_USERNAME and BSKY_APP_PASSWORD variables automately
USERNAME = BSKY_USERNAME
APP_PASSWORD = BSKY_APP_PASSWORD

# Authenticate steps:
client = Client()
client.login(USERNAME, APP_PASSWORD)

ProfileViewDetailed(did='did:plc:vtc4ozhahjwdh7fea5fjxpcv', handle='waterbottledent.bsky.social', associated=ProfileAssociated(activity_subscription=ProfileAssociatedActivitySubscription(allow_subscriptions='followers', py_type='app.bsky.actor.defs#profileAssociatedActivitySubscription'), chat=None, feedgens=0, labeler=False, lists=0, starter_packs=0, py_type='app.bsky.actor.defs#profileAssociated'), avatar='https://cdn.bsky.app/img/avatar/plain/did:plc:vtc4ozhahjwdh7fea5fjxpcv/bafkreidp2ti3dvuriz3kthi42wi4m5qp5micffcwboybfwtzg275gvvisu@jpeg', banner=None, created_at='2026-01-13T23:06:05.925Z', debug=None, description=None, display_name='', followers_count=2, follows_count=5, indexed_at='2026-01-13T23:08:03.625Z', joined_via_starter_pack=None, labels=[], pinned_post=None, posts_count=3, pronouns=None, status=None, verification=None, viewer=ViewerState(activity_subscription=None, blocked_by=False, blocking=None, blocking_by_list=None, followed_by=None, following=None, known_followers=Kn

In [86]:
# Read the starter packs dataset provided by Martin as a list of SP uri to gather data about the accounts within starter packs
df = pd.read_json("starterpacks.jsonl", lines=True)
df

list  \
0       at://did:plc:zdtz65xortdlxi7d6hlyz2j5/app.bsky...   
1       at://did:plc:zdtz65xortdlxi7d6hlyz2j5/app.bsky...   
2       at://did:plc:oextljnuf4ix335o7aapym55/app.bsky...   
3       at://did:plc:anumzyo4b5gclvho6uqkrpap/app.bsky...   
4       at://did:plc:d42nr7dwbfh4vfvduimuk7j5/app.bsky...   
...                                                   ...   
354690  at://did:plc:uos7xqdk7ggroepqnlzl7zev/app.bsky...   
354691  at://did:plc:huo5yychqhf6ifzoxek7bs4y/app.bsky...   
354692  at://did:plc:n2noxvecqcig4lvhthwcyp3q/app.bsky...   
354693  at://did:plc:gaeodbykn5riavylcovou3mh/app.bsky...   
354694  at://did:plc:fn4jlhgzfqrjtmpubpnf5gx3/app.bsky...   

                                             name                       $type  \
0       @‪motesandbeams.bsky.soci…'s Starter Pack  app.bsky.graph.starterpack   
1       @‪motesandbeams.bsky.soci…'s Starter Pack  app.bsky.graph.starterpack   
2                                       Knowledge  app.bsky.graph.starterpack   
3           ‪rngmom03.bsky.social‬'s Starter Pack  app.bsky.graph.starterpack   
4             ‪pietve.bsky.social‬'s Starter Pack  app.bsky.graph.starterpack   
...                                           ...                         ...   
354690   ‪rockmetal68.bsky.social‬'s Starter Pack  app.bsky.graph.starterpack   
354691  Cen - Creative & Candid �…'s Starter Pack  app.bsky.graph.starterpack   
354692  ‪mazzystar906.bsky.social‬'s Starter Pack  app.bsky.graph.starterpack   
354693               Ian Greenberg's Starter Pack  app.bsky.graph.starterpack   
354694      @‪cyndi52.bsky.social‬'s Starter Pack  app.bsky.graph.starterpack   

                       createdAt  \
0       2024-11-16T18:42:13.188Z   
1       2024-11-16T18:49:24.216Z   
2       2024-12-21T16:45:49.695Z   
3       2024-11-21T18:32:11.874Z   
4       2024-11-16T11:22:54.573Z   
...                          ...   
354690  2025-05-12T21:50:04.900Z   
354691  2025-05-14T12:31:52.439Z   
354692  2025-01-08T01:29:23.125Z   
354693  2025-01-12T21:17:20.652Z   
354694  2024-11-16T13:27:00.364Z   

                                                      cid  \
0       bafyreie3wdnxditj3sbinyctdum23xk3x6luvdy6voy3o...   
1       bafyreihb7o3slrv5i4uzl3uarlpb4gu2rluqsk6wzxhzv...   
2       bafyreidb7dfysn45jtudxaro6n6alqiemb5irmv2ztoan...   
3       bafyreibbdck4dxrmulfafmwai2rul3co52zqedrsxiuad...   
4       bafyreicggizjorcm6jldsbdrnlysxmwqgij4ywtvd7ljm...   
...                                                   ...   
354690  bafyreiecae6zmpsrvn4aspjdccmidq6byxapydls5zw24...   
354691  bafyreihty74mzmeubzcbuj4idgvbnsubwm26rj344wof3...   
354692  bafyreibfc3btykpkhyou74pdto4pv47lrbo4mmvr2zsr6...   
354693  bafyreidkco5kixpxxkemx3exwuynniprwt4f7ayjwl2x6...   
354694  bafyreie25o2fsthmkphspodlme5csg5u4ehxxkhuwjz47...   

                                  author  \
0       did:plc:zdtz65xortdlxi7d6hlyz2j5   
1       did:plc:zdtz65xortdlxi7d6hlyz2j5   
2       did:plc:oextljnuf4ix335o7aapym55   
3       did:plc:anumzyo4b5gclvho6uqkrpap   
4       did:plc:d42nr7dwbfh4vfvduimuk7j5   
...                                  ...   
354690  did:plc:uos7xqdk7ggroepqnlzl7zev   
354691  did:plc:huo5yychqhf6ifzoxek7bs4y   
354692  did:plc:n2noxvecqcig4lvhthwcyp3q   
354693  did:plc:gaeodbykn5riavylcovou3mh   
354694  did:plc:fn4jlhgzfqrjtmpubpnf5gx3   

                                                      uri           rkey  \
0       at://did:plc:zdtz65xortdlxi7d6hlyz2j5/app.bsky...  3lb3k7fkymr2f   
1       at://did:plc:zdtz65xortdlxi7d6hlyz2j5/app.bsky...  3lb3kmamw462k   
2       at://did:plc:oextljnuf4ix335o7aapym55/app.bsky...  3ldtdzitbxs2z   
3       at://did:plc:anumzyo4b5gclvho6uqkrpap/app.bsky...  3lbi3y3hdgx23   
4       at://did:plc:d42nr7dwbfh4vfvduimuk7j5/app.bsky...  3lb2rnu76p42x   
...                                                   ...            ...   
354690  at://did:plc:uos7xqdk7ggroepqnlzl7zev/app.bsky...  3loyxac7ew325   
354691  at://did:plc:huo5yychqhf6ifzoxek7bs4

### RQ1: Which account appear most in BlueSky Starter Packs?

Bluesky starter packs are curated lists of recommended accounts meant to help new users quickly find communities and high-quality content upon joining. Starter packs are created to mitigate "cold start" and were responsible for up to 43% of daily follow operations at their peak. Because each pack is created by a different user and focuses on a different theme or interest group, the accounts that appear most frequently across many starter packs are likely to be:

- broadly influential,

- highly visible across communities,

- Central hubs in the network.

Identifying these frequently-included accounts helps us understand:

- What kinds of voices are most prominent on Bluesky,

- Which users cross community boundaries,

- Whether certain media outlets, journalists, organizations, or personalities act as “anchor nodes” in the platform’s social ecosystem.

In [87]:
def get_accounts_info(account_list: List) -> List[Dict]:
    """
    Takes a list of account objects (each one is one entry from a starter pack)
    and extracts just the important identity information we care about.

    Parameters
    ----------
    account_list : list
        A list of accounts. Each account contains a 'subject' field with
        information about the actual user (like DID and handle).

    Returns
    -------
    list[dict]
        A list of simple dictionaries. Each dictionary has:
        - "Account DID": the unique ID for the account
        - "Account Handler": the user's handle (e.g. 'nytimes.com')
    """
    list_accounts = []
    account_dict = {}
    for account in account_list:
        # Each "account" is actually a dictionary-like structure
        # pulled from the starter pack. Inside it, "subject" stores
        # the actual user profile. We extract the two fields we care about.
        account_did = account["did"]    #old code: account["subject"]["did"]
        account_handler = account["handle"]    #old code: account["subject"]["handle"]
        account_dict = {"Account DID": account_did,
                        "Account Handler": account_handler}
        list_accounts.append(account_dict)
    return list_accounts

In [88]:
def get_starter_pack_info(starter_pack) -> dict:
    starter_pack_dict = {"SP DID": starter_pack["starter_pack"]["cid"],                            # unique ID for this starter pack record
                        "SP Creator DID": starter_pack["starter_pack"]["creator"]["did"],          # DID of the user who created the pack
                        "SP Creator Handle": starter_pack["starter_pack"]["creator"]["handle"],    # their handle (username)
                        "SP Description": starter_pack["starter_pack"]["record"]["description"]}   # text description of the pack}
    return starter_pack_dict

In [89]:
def get_feed_info(starting_feed_list: List) -> List[Dict]:
    feed_list = []
    for feed in starting_feed_list:
        feed_list.append({"Feed DID": feed["cid"],
                        "Feed CID": feed["did"],
                        "Feed Description": feed["description"], 
                        "Feed Creator DID": feed["creator"]["did"],
                        "Feed Creator Handle": feed["creator"]["handle"],
                        "Feed Like Count": feed["like_count"]})
    return feed_list

In [90]:
# def get_all_accounts(list_uri: str) -> List[Dict]:
#     accounts = []
#     cursor = None

#     while True:
#         

#         res = client.app.bsky.graph.get_list(params= {"list":list_uri,"limit":100})

        
#         accounts.append(res["items"])

#         if cursor is not None:
#             cursor = cursor
#         cursor = res["cursor"]
#         if not cursor:
#             break

#     return accounts

In [91]:
from typing import List, Dict, Any

def get_all_accounts(list_uri: str) -> List[Any]:
    """
    Fetch ALL accounts in a list (app.bsky.graph.list) and return a flat list
    of profile objects (subjects).

    Each returned element is either:
    - a dict with keys like "did", "handle", ...
    - or a ProfileView/ProfileViewDetailed object from atproto_client.
    """
    accounts: List[Any] = []
    cursor = None

    while True:
        params = {"list": list_uri, "limit": 100}
        if cursor:
            params["cursor"] = cursor

        res = client.app.bsky.graph.get_list(params=params)

        # Handle both dict-like and model-style responses
        #items = res.get('items') if isinstance(res, dict) else res.items

        # if res has attribute 'items' store the info, if not, return an empty list
        items = res.items if hasattr(res, 'items') else res.get('items',[])

        #print(items)
        
        if not items:
            break
        
        for item in items:
            # item is a ListItemView or dict; its "subject" is the account
            #if isinstance(item, dict):
                #subject = item.get("subject")  # this is throwing ModelNotFoundError for 'subject'
            #else:
                #subject = item.subject

            # check if subject is an attribute in the item and returns subject (account info),
            # if not, return items and append info
            subject = getattr(item, 'subject', item)
            accounts.append(subject)
        
        cursor = res.get("cursor") if isinstance(res, dict) else res.cursor

        #print(accounts)
        if not cursor:
            break

    return accounts

In [92]:
#@ return -> list[list[dict]]
def get_starter_packs_info(uris: List) -> List[List[Dict]]:
    """
    Given a list of starter pack URIs, download information about each starter pack
    and the (sample of) accounts included in it.

    For each starter pack URI, we:
      1. Call the Bluesky API to get a detailed view of that starter pack.
      2. Extract metadata about the starter pack (who created it, description, etc.).
      3. Extract a sample list of accounts that appear in that starter pack.
      4. Flatten this into one row per (starter pack, account) pair.

    Parameters
    ----------
    uris : list
        A list of starter pack AT-URIs (strings). Each URI identifies one starter pack.

    Returns
    -------
    list[dict]
        A list of dictionaries. Each dictionary is one row linking:
        - a specific starter pack
        - one account that appears in that pack (from the sample list)
    """
    # this will hold all rows across all starter packs
    starter_packs_list = []
    feed_list = []

    sb_dict = {}

    for uri in uris:
        try:
            # 1. Ask the Bluesky API for detailed information about this starter pack
            starter_pack = client.app.bsky.graph.get_starter_pack(params={"starterPack": uri})
        except exceptions.BadRequestError as e:
            # If the server says "starter pack not found", we skip this URI and continue.
            print("Skipping URI (starter pack not found):", uri)
            continue
        
        # 2) Extract some basic metadata about this starter pack with get_starter_pack_info()
        starter_pack_info = get_starter_pack_info(starter_pack)
        
        # 3) Get the sample of accounts that appear in this starter pack
        #unproccessed_account_list_from_sb = starter_pack["starter_pack"]["list_items_sample"]
        # pass list of account uris from sp
        unproccessed_account_list_from_sb = get_all_accounts(starter_pack["starter_pack"]["list"]["uri"])

        #print(unproccessed_account_list_from_sb)
        
        # Use our helper to extract just DID + handle for each account in the sample
        proccessed_list = get_accounts_info(unproccessed_account_list_from_sb)

        if len(starter_pack["starter_pack"]["feeds"]) != 0:
            feed_info = get_feed_info(starter_pack["starter_pack"]["feeds"])
            for feed in feed_info:
                feed_list.append({"SP URI": uri,
                        "SP DID": starter_pack_info["SP DID"],
                        "SP Creator DID": starter_pack_info["SP Creator DID"],
                        "SP Creator Handle": starter_pack_info["SP Creator Handle"],
                        "SP Description": starter_pack_info["SP Description"],
                        "Feed DID": feed["Feed DID"],
                        "Feed CID": feed["Feed CID"],
                        "Feed Description": feed["Feed Description"], 
                        "Feed Creator DID": feed["Feed Creator DID"],
                        "Feed Creator Handle": feed["Feed Creator Handle"],
                        "Feed Like Count": feed["Feed Like Count"]
                        })

        # 4) For each account in this starter pack's sample, create one flat row
        for account in proccessed_list:
            sb_dict = {"SP URI": uri,
                        "SP DID": starter_pack_info["SP DID"],
                        "SP Creator DID": starter_pack_info["SP Creator DID"],
                        "SP Creator Handle": starter_pack_info["SP Creator Handle"],
                        "SP Description": starter_pack_info["SP Description"],
                        "AccountDID": account["Account DID"],
                        "Account Handler": account["Account Handler"]}
            starter_packs_list.append(sb_dict)
    
    return [starter_packs_list,feed_list]

### Creating Testing Dataset

In [93]:
test_100 = df[:100]
test_1000 = df[:1000]

test_100.head()

list  \
0  at://did:plc:zdtz65xortdlxi7d6hlyz2j5/app.bsky...   
1  at://did:plc:zdtz65xortdlxi7d6hlyz2j5/app.bsky...   
2  at://did:plc:oextljnuf4ix335o7aapym55/app.bsky...   
3  at://did:plc:anumzyo4b5gclvho6uqkrpap/app.bsky...   
4  at://did:plc:d42nr7dwbfh4vfvduimuk7j5/app.bsky...   

                                        name                       $type  \
0  @‪motesandbeams.bsky.soci…'s Starter Pack  app.bsky.graph.starterpack   
1  @‪motesandbeams.bsky.soci…'s Starter Pack  app.bsky.graph.starterpack   
2                                  Knowledge  app.bsky.graph.starterpack   
3      ‪rngmom03.bsky.social‬'s Starter Pack  app.bsky.graph.starterpack   
4        ‪pietve.bsky.social‬'s Starter Pack  app.bsky.graph.starterpack   

                  createdAt  \
0  2024-11-16T18:42:13.188Z   
1  2024-11-16T18:49:24.216Z   
2  2024-12-21T16:45:49.695Z   
3  2024-11-21T18:32:11.874Z   
4  2024-11-16T11:22:54.573Z   

                                                 cid  \
0  bafyreie3wdnxditj3sbinyctdum23xk3x6luvdy6voy3o...   
1  bafyreihb7o3slrv5i4uzl3uarlpb4gu2rluqsk6wzxhzv...   
2  bafyreidb7dfysn45jtudxaro6n6alqiemb5irmv2ztoan...   
3  bafyreibbdck4dxrmulfafmwai2rul3co52zqedrsxiuad...   
4  bafyreicggizjorcm6jldsbdrnlysxmwqgij4ywtvd7ljm...   

                             author  \
0  did:plc:zdtz65xortdlxi7d6hlyz2j5   
1  did:plc:zdtz65xortdlxi7d6hlyz2j5   
2  did:plc:oextljnuf4ix335o7aapym55   
3  did:plc:anumzyo4b5gclvho6uqkrpap   
4  did:plc:d42nr7dwbfh4vfvduimuk7j5   

                                                 uri           rkey  \
0  at://did:plc:zdtz65xortdlxi7d6hlyz2j5/app.bsky...  3lb3k7fkymr2f   
1  at://did:plc:zdtz65xortdlxi7d6hlyz2j5/app.bsky...  3lb3kmamw462k   
2  at://did:plc:oextljnuf4ix335o7aapym55/app.bsky...  3ldtdzitbxs2z   
3  at://did:plc:anumzyo4b5gclvho6uqkrpap/app.bsky...  3lbi3y3hdgx23   
4  at://did:plc:d42nr7dwbfh4vfvduimuk7j5/app.bsky...  3lb2rnu76p42x   

                   collection_time  \
0 2026-02-11 08:00:03.164000+00:00   
1 2026-02-11 08:00:03.164000+00:00   
2 2026-02-11 08:00:17.963000+00:00   
3 2026-02-11 08:00:24.338000+00:00   
4 2026-02-11 08:00:25.661000+00:00   

                                               feeds  \
0                                                NaN   
1                                                NaN   
2                                                 []   
3  [{'uri': 'at://did:plc:z72i7hdynmk6r22z27h6tvu...   
4  [{'uri': 'at://did:plc:5rw2on4i56btlcajojaxwca...   

                  updatedAt  \
0                       NaN   
1                       NaN   
2  2025-01-09T12:32:19.214Z   
3                       NaN   
4                       NaN   

                                         description descriptionFacets image  
0                                                NaN               NaN   NaN  
1                                                NaN               NaN   NaN  
2  Education is the most powerful weapon which yo...               NaN   NaN  
3                                                NaN               NaN   NaN  
4                                                NaN               NaN   NaN

In [94]:
# testing uri
test_100.iloc[1,6]


'at://did:plc:zdtz65xortdlxi7d6hlyz2j5/app.bsky.graph.starterpack/3lb3kmamw462k'

### Run the Script to Get Starter Packs Accounts and Starter Packs Feeds Dataset

In [95]:
result = get_starter_packs_info(test_100['uri'])
accounts_df = pd.DataFrame(result[0])
feeds_df = pd.DataFrame(result[1])


### Top 10 Accounts Appear Most Frequenly within Starter Packs

In [96]:
#Count number of appearances within starter packs and store the result into DataFrame
account_appearance_count = accounts_df["Account Handler"].value_counts()
df_counts = account_appearance_count.reset_index()
df_counts.columns = ["Account Handler", "Count"]

# Add frequencies column
total = account_appearance_count.sum()
df_counts["Frequency"] = df_counts["Count"] / total
print(len(df_counts))
df_counts.head(10)

6185


Account Handler  Count  Frequency
0  charlietotherescue.bsky.social     49   0.005874
1                        bsky.app     27   0.003237
2        oh-its-chris.bsky.social     27   0.003237
3          donnacrowe.bsky.social     27   0.003237
4         evelynelotz.bsky.social     25   0.002997
5            dannyj73.bsky.social     22   0.002637
6          tamiguitar.bsky.social     22   0.002637
7     friendsofbcdogs.bsky.social     21   0.002517
8             pgmpinc.bsky.social     20   0.002398
9      mtncannoligirl.bsky.social     18   0.002158

### Result: Most Frequently Appearing Accounts

The accounts listed above appear in the highest number of starter packs in our dataset. This indicates that these accounts are repeatedly recommended by different starter pack creators, rather than appearing randomly or uniformly across all packs.

This supports our hypothesis that certain accounts are **promoted more frequently, suggesting that visibility on Bluesky may be concentrated among a smaller subset of accounts.** Because starter packs function as curated discovery tools, accounts that appear more often are more likely to be discovered and followed by new users. This repeated inclusion increases their visibility and reinforces their presence within the platform’s social network. 

This is similar to the principle of the "rich gets richer," also known as preferential attachment, where accounts that already appear frequently in starter packs become even more likely to be included in additional starter packs. As these accounts gain more exposure through repeated inclusion, they become more visible to users and are more likely to be followed, recognized, and recommended by others. This creates a positive feedback loop in which popular accounts continue to accumulate visibility at a faster rate than less frequently included accounts.

This finding supports our hypothesis that starter packs play a role in amplifying certain accounts and suggests that discovery on Bluesky may be influenced by cumulative visibility effects rather than purely decentralized exploration.

In [97]:
#TODO: Right now the result only include partial of the accounts included within the bluesky. 
# Need to get all to add into the dataset

### RQ2: What feeds appear the most among all starter packs with feeds?

In [98]:
#Count number of appearances within starter packs and store the result into DataFrame
feeds_appearance_count = feeds_df["Feed Creator Handle"].value_counts()
df_counts = feeds_appearance_count.reset_index()
df_counts.columns = ["Feed Creator Handle", "Count"]

# Add frequencies column
total = feeds_appearance_count.sum()
df_counts["Frequency"] = df_counts["Count"] / total
df_counts.head(10)

Feed Creator Handle  Count  Frequency
0                   bsky.app      5    0.15625
1                skyfeed.xyz      4    0.12500
2  colinbaines15.bsky.social      3    0.09375
3           eepy.bsky.social      2    0.06250
4        timmersionmedia.com      2    0.06250
5              shultzman.com      2    0.06250
6             clarabelle.xyz      2    0.06250
7                 aendra.com      2    0.06250
8                   bsky.art      2    0.06250
9             bossett.social      1    0.03125

Similarly to our finding above, Bluesky still has the highest percentage of feeds shared by starter packs within our test data. Although the count difference is far less than Bluesky's appearance as an account shared by starter packs, it is still apparent. Once again, we only sampled a small subset of the data and need to further look into the rest of the dataset.

# Question 3: Thematic Categorization

Starter packs on Bluesky often reflecting communities built around specific interests such as politics, gaming, news,art, or travel. To better understand the thematic structure of our dataset, we developed a simple rule-based text classification method that assigns each starter pack to a category based on the keywords found in its name and description. This allow us to:
- Quantify topics and theme to see which are most common across the starter packs.
- Enable meaningful approximation that enable downstream quantitative analysis such as frequency counts 
- Comparison across categories

In [99]:
import re
from typing import Dict, List

# 1. Define the set of possible categories

STARTER_PACK_CATEGORIES = [
    "gaming", "sports", "music", "politics", "science", "tech", "finance",
    "art", "fashion", "food", "travel", "education", "health", "books",
    "movies", "animals", "religion", "nature", "comedy", "news", "unknown"
]

# 2. Keyword lists for each category
# A starter pack will be classified by matching
# these keywords to its name/description text.


STARTER_PACK_KEYWORDS: Dict[str, List[str]] = {
    "gaming":   ["game", "gaming", "pc", "console", "xbox", "playstation", "steam", "fps", "rpg", "minecraft", "valorant"],
    "sports":   ["sport", "soccer", "football", "basketball", "baseball", "tennis", "golf", "nba", "nfl", "fifa"],
    "music":    ["music", "band", "album", "song", "playlist", "dj", "concert", "guitar", "piano", "rapper"],
    "politics": ["politic", "election", "vote", "policy", "senate", "congress", "government", "campaign", "mayor","Miniter","mps","diplomacy"],
    "science":  ["science", "biology", "chemistry", "physics", "genetics", "lab", "research", "paper", "journal"],
    "tech":     ["tech", "software", "developer", "coding", "python", "javascript", "ai", "ml", "data", "cloud", "startup"],
    "finance":  ["finance", "invest", "stock", "trading", "crypto", "bitcoin", "portfolio", "economy", "market"],
    "art":      ["art", "artist", "drawing", "painting", "illustration", "sketch", "gallery", "design"],
    "fashion":  ["fashion", "style", "outfit", "streetwear", "runway", "vogue", "makeup", "beauty"],
    "food":     ["food", "cook", "recipe", "baking", "restaurant", "coffee", "tea", "chef"],
    "travel":   ["travel", "trip", "vacation", "flight", "hotel", "backpack", "itinerary", "tour"],
    "education":["education", "school", "university", "class", "homework", "study", "course", "lecture"],
    "health":   ["health", "fitness", "workout", "gym", "diet", "nutrition", "doctor", "medicine", "therapy"],
    "books":    ["book", "novel", "reading", "literature", "author", "kindle"],
    "movies":   ["movie", "film", "cinema", "director", "actor", "netflix", "tv", "series"],
    "animals":  ["animal", "cat", "dog", "pet", "wildlife", "bird", "fish"],
    "religion": ["religion", "faith", "church", "bible", "islam", "quran", "hindu", "buddha"],
    "nature":   ["nature", "outdoors", "hiking", "camping", "climate", "forest", "mountain", "ocean"],
    "comedy":   ["comedy", "joke", "meme", "funny", "satire"],
    "news":     ["news", "journalism", "reporter", "breaking", "headline"],
}


# 3) Helper: Normalize text into tokens
# Converts text -> lowercase -> word tokens

_WORD_RE = re.compile(r"[a-z0-9]+")

def _normalize(text: str) -> List[str]:
   
    """
    Normalize text and finds matching words to the category
    """
    
    text = (text or "").lower()
    return _WORD_RE.findall(text)
    
# 4. Helper: Score every category
# Counts how many keywords from each category appear in the text.
def _score_categories(tokens: List[str]) -> Dict[str, int]:
    """
    This takes each token and puts them into dictonary. 
    Return the score of the token. 
    """
    scores = {cat: 0 for cat in STARTER_PACK_KEYWORDS.keys()}
    joined = " ".join(tokens)
    for cat, kws in STARTER_PACK_KEYWORDS.items():
        for kw in kws:
            if kw in joined:
                scores[cat] += 2
            else:
                scores[cat] += sum(1 for t in tokens if t == kw)
    return scores

# 5. Main classifier function
def classify_starter_pack(name: str, description: str) -> str:

    """
    Takes in the name and descrption of starter pack. 
    Returns the Themes of the starter pack

    """

    
    text = description or ''
    tokens = _normalize(text)

    scores = _score_categories(tokens)
    best_cat = max(scores, key=scores.get)

    return best_cat if scores[best_cat] >= 2 else "Unknown"

In [100]:
df= test_1000.loc[:, ['name', 'description']]
df2 = df.copy()

In [101]:
# 6) Apply classifier to the dataset
df2["name"] = df2["name"].fillna("").astype(str)
df2["description"] = df2["description"].fillna("").astype(str)

df2 = df2.drop_duplicates(subset=["description"])

df2["Starter Pack Classification"] = df2.apply(
    lambda row: classify_starter_pack(row["name"], row["description"]),
    axis=1
)

df2.head()

name  \
0           @‪motesandbeams.bsky.soci…'s Starter Pack   
2                                           Knowledge   
7                                     We Love Memoirs   
11  UK MPs & peers working for better UK-EU relations   
14                        Greater Manchester politics   

                                          description  \
0                                                       
2   Education is the most powerful weapon which yo...   
7          For memoir authors and readers to connect!   
11  Nearly 100 MPs, peers, ex-MPs or ex-MEPs on Bl...   
14  MPs, councillors and others politicos in Great...   

   Starter Pack Classification  
0                      Unknown  
2                    education  
7                        books  
11                    politics  
14                    politics

In [102]:
df2["Starter Pack Classification"].value_counts()

Starter Pack Classification
Unknown      132
art           61
animals       54
tech          30
politics      14
science       12
gaming        11
music          9
sports         6
news           5
education      5
food           5
books          4
comedy         4
health         4
movies         2
finance        1
Name: count, dtype: int64

Similarly to our finding above, Bluesky still has the highest percentage of feeds shared by starter packs within our test data. Although the count difference is far less than Bluesky's appearance as an account shared by starter packs, it is still apparent. Once again, we only sampled a small subset of the data and need to further look into the rest of the dataset.

### RQ3: Thematic Categorization Analysis

There were 132 starter packs without descriptions, and the top five themes identified were art, animals, tech, politics, and science. 

#### **Uncertainties**: 
There might be misclassification of themes based on their given descriptions and titles. 
Themes may be too broad, especially considering that categories like art include many different mediums (ex. Film, drawing/painting, calligraphy, etc.)

##### **New Problem**: 
To define echo chambers, we will need to determine the percentage of overlapping followers within a starter pack theme. This will increase the data size  of the data and computational cost. 

#### **Next steps**: 
- We will continue collecting starter pack data and increase the dataset size. To identify echo chambers, we will randomly pick 5 starter packs (test_data) to prevent exceeding our API limit. We will then calculate followers overlaps within narrowed popular themes and split the data processing among team members to reduce the computational load.
- We can also revise the categorzing algorithm and use LLM instead of rule-based
- We can also fetch the feeds and use it material for categorizing starter packs as an additional to name and description
